File: src/test/java/com/mercateo/jsonschema/property/PropertyBuilderTest.java
package com.mercateo.jsonschema.property;

import javaslang.collection.Iterator;
import javaslang.collection.List;
import javaslang.collection.Set;
import javaslang.collection.Traversable;
import org.junit.Before;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.annotation.Annotation;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

public class PropertyBuilderTest {
    private Logger log = LoggerFactory.getLogger(PropertyBuilderTest.class);

    private PropertyBuilder propertyBuilder;

    public static <T> T getFirstElement(Traversable<T> collection) {
        return collection.iterator().next();
    }

    @Before
    public void setUp() throws Exception {
        propertyBuilder = new PropertyBuilderDefault(List.of(new FieldCollector(FieldCollectorConfig.builder().build()), new MethodCollector()));
    }

    @Test
    public void rootPropertyDefaultNameIsHashCharacter() {
        Property property = propertyBuilder.from(PropertyHolder.class);

        assertThat(property.name()).isEqualTo("#");
    }

    @Test
    public void callingValueAccessorOfRootElementThrows() {
        Property property = propertyBuilder.from(PropertyHolder.class);

        assertThatThrownBy(() -> property.getValue(null)).isInstanceOf(IllegalStateException.class)
                .hasMessage("cannot call value accessor for root element");
    }

    @Test
    public void containsChildFromField() throws Exception {
        Property property = propertyBuilder.from(PropertyHolder.class);

        assertThat(property.children()).extracting(Property::name).containsExactly("property");
    }

    @Test
    public void buildsPropertiesRecursively() throws Exception {
        Property property = propertyBuilder.from(TwoLevelPropertyHolder.class);

        final List<Property> children = property.children();
        final Property firstLevelElement = getFirstElement(children);
        final Property secondLevelElement = getFirstElement(firstLevelElement.children());

        assertThat(secondLevelElement.name()).isEqualTo("property");
    }

    @Test
    public void createsPropertyWithGenerics() throws Exception {

        Property property = propertyBuilder.from(TwoLevelPropertyHolder.class);

        final Property firstElement = getFirstElement(property.children());
        final Property secondLevelElement = getFirstElement(firstElement.children());
        assertThat(secondLevelElement.genericType().getRawType()).isEqualTo(String.class);
    }

    @Test
    public void propertyReturnsPropertyValue() throws Exception {
        Property property = propertyBuilder.from(PropertyHolder.class);

        final Property firstElement = getFirstElement(property.children());

        final PropertyHolder propertyHolder = new PropertyHolder();
        propertyHolder.property = "foo";

        assertThat(firstElement.getValue(propertyHolder)).isEqualTo("foo");
    }

    @Test
    public void propertyReturnsClassAnnotation() throws Exception {
        Property property = propertyBuilder.from(PropertyHolder.class);

        final Annotation firstElement = getFirstElement(property.annotations().head()._2());
        assertThat(firstElement).isInstanceOf(Annotation1.class);
    }

    @Test
    public void subPropertyReturnsFieldAnnotation() throws Exception {
        Property property = propertyBuilder.from(TwoLevelPropertyHolder.class);

        final Property firstElement1 = getFirstElement(property.children());
        assertThat(getAnnotations(firstElement1).toJavaSet()).containsExactlyInAnyOrder(Annotation1.class, Annotation2.class);
    }

    private Set<Class<? extends Annotation>> getAnnotations(Property firstElement1) {
        return firstElement1.annotations().keySet();
    }

    @Test
    public void returnInheritedProperty() throws Exception {
        Property property = propertyBuilder.from(InheritedPropertyHolder.class);

        assertThat(property.children()).extracting(Property::name).containsExactly("property");
    }

    @Test
    public void nonObjectTypesShouldHaveNoChildren() throws Exception {
        Property property = propertyBuilder.from(PropertyHolder.class);

        final Property firstElement = getFirstElement(property.children());
        assertThat(firstElement.children()).isEmpty();
    }

    @Test
    public void returnsIdentialTypeDescriptorsForSameType() throws Exception {
        Property property1 = propertyBuilder.from(PropertyHolder.class);
        Property property2 = propertyBuilder.from(PropertyHolder.class);

        assertThat(property1).isNotSameAs(property2);
        assertThat(property1.propertyDescriptor()).isSameAs(property2.propertyDescriptor());
    }

    @Test
    public void returnMethodProperty() throws Exception {
        Property property = propertyBuilder.from(MethodPropertyHolder.class);
        final Property firstElement = getFirstElement(property.children());
        assertThat(firstElement.name()).isEqualTo("property");
    }

    @Test
    public void returnMethodPropertyValue() throws Exception {
        Property property = propertyBuilder.from(MethodPropertyHolder.class);
        final Property firstElement = getFirstElement(property.children());

        final MethodPropertyHolder methodPropertyHolder = new MethodPropertyHolder();
        assertThat(firstElement.getValue(methodPropertyHolder)).isEqualTo("foo");
    }

    @Test
    public void returnMethodAnnotations() throws Exception {
        Property property = propertyBuilder.from(MethodPropertyHolder.class);
        final Property firstElement = getFirstElement(property.children());

        assertThat(getAnnotations(firstElement)).containsExactly(Annotation2.class);
    }

    @Test
    public void returnMethodAndClassAnnotations() throws Exception {
        Property property = propertyBuilder.from(TwoLevelMethodPropertyHolder.class);
        final Property firstElement = getFirstElement(property.children());

        assertThat(getAnnotations(firstElement)).containsExactlyInAnyOrder(Annotation1.class, Annotation3.class);
    }

    @Test
    public void followCollectionProperties() throws Exception {
        Property property = propertyBuilder.from(CollectionPropertyHolder.class);
        final Property collectionElement = getFirstElement(property.children());
        final Property collectionTypeElement = getFirstElement(collectionElement.children());

        assertThat(collectionTypeElement).isNotNull();
        assertThat(collectionTypeElement.name()).isEqualTo("");
        assertThat(collectionTypeElement.annotations().values()).isEmpty();
        assertThat(collectionTypeElement.genericType().getRawType()).isEqualTo(String.class);

        final CollectionPropertyHolder collectionPropertyHolder = new CollectionPropertyHolder();
        collectionPropertyHolder.values = List.of("foo");

        assertThat(collectionTypeElement.getValue(collectionPropertyHolder)).isNull();
    }

    @Test
    public void followNestedCollectionProperties() throws Exception {
        Property property = propertyBuilder.from(NestedCollectionPropertyHolder.class);
        final Property collectionElement = getFirstElement(property.children());

        final Property collectionTypeElement = getFirstElement(collectionElement.children());
        assertThat(collectionTypeElement).isNotNull();
        assertThat(collectionTypeElement.name()).isEqualTo("");
        assertThat(collectionTypeElement.annotations().values()).isEmpty();
        assertThat(collectionTypeElement.genericType().getRawType()).isEqualTo(String[].class);

        final Property nestedCollectionTypeElement = getFirstElement(collectionTypeElement.children());

        assertThat(nestedCollectionTypeElement).isNotNull();
        assertThat(nestedCollectionTypeElement.name()).isEqualTo("");
        assertThat(nestedCollectionTypeElement.genericType().getRawType()).isEqualTo(String.class);
        assertThat(nestedCollectionTypeElement.annotations().values()).isEmpty();
    }

    @Test
    public void returnPropertyReferenceIfRecursive() throws Exception {
        Property property = propertyBuilder.from(RecursivePropertyHolder.class);
        final Iterator<Property> childrenIterator = property.children().iterator();
        final Property firstElement = childrenIterator.next();

        assertThat(firstElement.name()).isEqualTo("children");

        final Property firstSubelement = getFirstElement(firstElement.children() );
        assertThat(firstSubelement.name()).isEmpty();
        assertThat(firstSubelement.genericType().getRawType()).isEqualTo(RecursivePropertyHolder.class);
        assertThat(firstSubelement.children()).isEmpty();
        final PropertyDescriptor propertyDescriptor = firstSubelement.propertyDescriptor();
        // TODO fix test
        //assertThat(propertyDescriptor.children()).hasSize(2);

        final Property secondElement = childrenIterator.next();
        assertThat(secondElement.name()).isEqualTo("name");
        assertThat(secondElement.genericType().getRawType()).isEqualTo(String.class);
    }

    @Retention(RetentionPolicy.RUNTIME)
    @interface Annotation1 {
    }

    @Retention(RetentionPolicy.RUNTIME)
    @interface Annotation2 {
    }

    @Retention(RetentionPolicy.RUNTIME)
    @interface Annotation3 {
    }

    @Annotation1
    static class PropertyHolder {
        String property;
    }

    @Annotation1
    static class MethodPropertyHolder {
        @Annotation2
        String getProperty() {
            return "foo";
        }
    }

    static class TwoLevelMethodPropertyHolder {
        @Annotation3
        MethodPropertyHolder holder;
    }

    static class TwoLevelPropertyHolder {
        @Annotation1
        GenericPropertyHolder<String> holder;
    }

    @Annotation2
    static class GenericPropertyHolder<T> {
        T property;
    }

    static class InheritedPropertyHolder extends PropertyHolder {
    }

    static class CollectionPropertyHolder {
        List<String> values;
    }

    static class NestedCollectionPropertyHolder {
        List<String[]> values;
    }

    static class RecursivePropertyHolder {
        List<RecursivePropertyHolder> children;

        String name;
    }
}

File: src/test/java/com/mercateo/jsonschema/property/PropertyBuilderTest.kt
package com.mercateo.jsonschema.property

import com.mercateo.jsonschema.property.PropertyBuilderClasses.*
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.IterableAssert
import org.junit.Before
import org.junit.Test
import org.slf4j.LoggerFactory

class PropertyBuilderTest {
    private val log = LoggerFactory.getLogger(PropertyBuilderTest::class.java)

    private lateinit var propertyBuilder: PropertyBuilder

    @Before
    fun setUp() {
        propertyBuilder = PropertyBuilderDefault(FieldCollector(), MethodCollector())
    }

    @Test
    fun rootPropertyDefaultNameIsHashCharacter() {
        val property = propertyBuilder.from(PropertyHolder::class.java)

        assertThat(property.name).isEqualTo("#")
    }

    @Test
    fun containsChildFromField() {
        val property = propertyBuilder.from(PropertyHolder::class.java)

        assertThat(property.children).extracting("name").containsExactly("property")
    }

    @Test
    fun containsEnumChildAsString() {
        val property = propertyBuilder.from(EnumPropertyHolder::class.java)

        assertThat(property.children).extracting("name").containsExactly("enumProperty")
        assertThat(property.children).extracting("propertyType").containsExactly(PropertyType.STRING)
    }

    @Test
    fun buildsPropertiesRecursively() {
        val property = propertyBuilder.from(TwoLevelPropertyHolder::class.java)

        val children = property.children
        val firstLevelElement = children.first()
        val secondLevelElement = firstLevelElement.children.first()

        assertThat(secondLevelElement.name).isEqualTo("property")
    }

    @Test
    fun createsPropertyWithGenerics() {

        val property = propertyBuilder.from(TwoLevelPropertyHolder::class.java)

        val firstElement = property.children.first()
        val secondLevelElement = firstElement.children.first()
        assertThat(secondLevelElement.genericType.rawType).isEqualTo(String::class.java)
    }

    @Test
    fun propertyReturnsPropertyValue() {
        val property = propertyBuilder.from(PropertyHolder::class.java)

        val firstElement = property.children.first()

        val propertyHolder = PropertyHolder()
        propertyHolder.property = "foo"

        assertThat(firstElement.getValue(propertyHolder)).isEqualTo("foo")
    }

    @Test
    fun propertyReturnsClassAnnotation() {
        val property = propertyBuilder.from(PropertyHolder::class.java)

        val firstElement = property.annotations.values.first().first()
        assertThat(firstElement).isInstanceOf(Annotation1::class.java)
    }

    @Test
    @Throws(Exception::class)
    fun subPropertyReturnsFieldAnnotation() {
        val property = propertyBuilder.from(TwoLevelPropertyHolder::class.java)

        val firstElement1 = property.children.first()
        assertThat(getAnnotations(firstElement1)).containsExactlyInAnyOrder(Annotation1::class.java, Annotation2::class.java)
    }

    fun <T> assertThat(actual: Iterable<T>) = IterableAssert(actual)

    private fun getAnnotations(firstElement1: Property<*, *>): Set<Class<out Annotation>> {
        return firstElement1.annotations.keys
    }

    @Test
    @Throws(Exception::class)
    fun returnInheritedProperty() {
        val property = propertyBuilder.from(PropertyBuilderClasses.InheritedPropertyHolder::class.java)

        assertThat(property.children).extracting("name").containsExactly("property")
    }

    @Test
    @Throws(Exception::class)
    fun nonObjectTypesShouldHaveNoChildren() {
        val property = propertyBuilder.from(PropertyHolder::class.java)

        val firstElement = property.children.first()
        assertThat(firstElement.children).isEmpty()
    }

    @Test
    @Throws(Exception::class)
    fun returnsIdentialTypeDescriptorsForSameType() {
        val property1 = propertyBuilder.from(PropertyHolder::class.java)
        val property2 = propertyBuilder.from(PropertyHolder::class.java)

        assertThat(property1).isNotSameAs(property2)
        assertThat(property1.propertyDescriptor).isSameAs(property2.propertyDescriptor)
    }

    @Test
    @Throws(Exception::class)
    fun returnMethodProperty() {
        val property = propertyBuilder.from(MethodPropertyHolder::class.java)
        val firstElement = property.children.first()
        assertThat(firstElement.name).isEqualTo("property")
    }

    @Test
    @Throws(Exception::class)
    fun returnMethodPropertyValue() {
        val property = propertyBuilder.from(MethodPropertyHolder::class.java)
        val firstElement = property.children.first()

        val methodPropertyHolder = MethodPropertyHolder()
        assertThat(firstElement.getValue(methodPropertyHolder)).isEqualTo("foo")
    }

    @Test
    @Throws(Exception::class)
    fun returnMethodAnnotations() {
        val property = propertyBuilder.from(MethodPropertyHolder::class.java)
        val firstElement = property.children.first()

        assertThat(getAnnotations(firstElement)).containsExactly(Annotation2::class.java)
    }

    @Test
    @Throws(Exception::class)
    fun returnMethodAndClassAnnotations() {
        val property = propertyBuilder.from(TwoLevelMethodPropertyHolder::class.java)
        val firstElement = property.children.first()

        assertThat(getAnnotations(firstElement)).containsExactlyInAnyOrder(Annotation1::class.java, Annotation3::class.java)
    }

    @Test
    @Throws(Exception::class)
    fun followCollectionProperties() {
        val property: Property<Void, CollectionPropertyHolder> = propertyBuilder.from(CollectionPropertyHolder::class.java)
        val collectionElement: Property<CollectionPropertyHolder, Any> = property.children.first()
        val collectionTypeElement: Property<Any, out Any?> = collectionElement.children.first()

        assertThat(collectionTypeElement).isNotNull()
        assertThat(collectionTypeElement.name).isEqualTo("")
        assertThat(collectionTypeElement.annotations.values).isEmpty()
        assertThat(collectionTypeElement.genericType.rawType).isEqualTo(String::class.java)

        val collectionPropertyHolder = CollectionPropertyHolder()
        collectionPropertyHolder.values = listOf("foo")

        assertThat(collectionTypeElement.getValue(collectionPropertyHolder)).isNull()
    }

    @Test
    @Throws(Exception::class)
    fun followNestedCollectionProperties() {
        val property = propertyBuilder.from(NestedCollectionPropertyHolder::class.java)
        val collectionElement = property.children.first()

        val collectionTypeElement = collectionElement.children.first()
        assertThat(collectionTypeElement).isNotNull()
        assertThat(collectionTypeElement.name).isEqualTo("")
        assertThat(collectionTypeElement.annotations.values).isEmpty()
        assertThat(collectionTypeElement.genericType.rawType).isEqualTo(Array<String>::class.java)

        val nestedCollectionTypeElement = collectionTypeElement.children.first()

        assertThat(nestedCollectionTypeElement).isNotNull()
        assertThat(nestedCollectionTypeElement.name).isEqualTo("")
        assertThat(nestedCollectionTypeElement.genericType.rawType).isEqualTo(String::class.java)
        assertThat(nestedCollectionTypeElement.annotations.values).isEmpty()
    }

    @Test
    @Throws(Exception::class)
    fun returnsPropertyWithoutChildrenIfRecursive() {
        val property = propertyBuilder.from(RecursivePropertyHolder::class.java)
        val childrenIterator = property.children.iterator()
        val firstElement = childrenIterator.next()

        assertThat(firstElement.name).isEqualTo("children")

        val firstSubelement = firstElement.children.first()
        assertThat(firstSubelement.name).isEmpty()
        assertThat(firstSubelement.genericType.rawType).isEqualTo(RecursivePropertyHolder::class.java)
        val propertyDescriptor = firstSubelement.propertyDescriptor
        assertThat(propertyDescriptor.children).isEmpty()

        val secondElement = childrenIterator.next()
        assertThat(secondElement.name).isEqualTo("name")
        assertThat(secondElement.genericType.rawType).isEqualTo(String::class.java)
    }

    @Test
    fun shouldMapStringType() {
        val property = propertyBuilder.from(TypesPropertyHolder::class.java)

        assertThat(property.children.find { it.name == "string" }).extracting("propertyType").contains(PropertyType.STRING)
    }

    @Test
    fun shouldMapPrimitiveBooleanType() {
        val property = propertyBuilder.from(TypesPropertyHolder::class.java)

        assertThat(property.children.find { it.name == "booleanPrimitive" }).extracting("propertyType").contains(PropertyType.BOOLEAN)
    }

    @Test
    fun shouldMapBooleanType() {
        val property = propertyBuilder.from(TypesPropertyHolder::class.java)

        assertThat(property.children.find { it.name == "booleanValue" }).extracting("propertyType").contains(PropertyType.BOOLEAN)
    }

    @Test
    fun shouldMapPrimitiveIntegerType() {
        val property = propertyBuilder.from(TypesPropertyHolder::class.java)

        assertThat(property.children.find { it.name == "integerPrimitive" }).extracting("propertyType").contains(PropertyType.INTEGER)
    }

    @Test
    fun shouldMapIntegerType() {
        val property = propertyBuilder.from(TypesPropertyHolder::class.java)

        assertThat(property.children.find { it.name == "integerValue" }).extracting("propertyType").contains(PropertyType.INTEGER)
    }

    @Test
    fun shouldMapPrimitiveLongType() {
        val property = propertyBuilder.from(TypesPropertyHolder::class.java)

        assertThat(property.children.find { it.name == "longPrimitive" }).extracting("propertyType").contains(PropertyType.INTEGER)
    }

    @Test
    fun shouldMapLongType() {
        val property = propertyBuilder.from(TypesPropertyHolder::class.java)

        assertThat(property.children.find { it.name == "longValue" }).extracting("propertyType").contains(PropertyType.INTEGER)
    }

    @Test
    fun shouldMapPrimitiveFloatType() {
        val property = propertyBuilder.from(TypesPropertyHolder::class.java)

        assertThat(property.children.find { it.name == "floatPrimitive" }).extracting("propertyType").contains(PropertyType.NUMBER)
    }

    @Test
    fun shouldMapFloatType() {
        val property = propertyBuilder.from(TypesPropertyHolder::class.java)

        assertThat(property.children.find { it.name == "floatValue" }).extracting("propertyType").contains(PropertyType.NUMBER)
    }


    @Test
    fun shouldMapPrimitiveDoubleType() {
        val property = propertyBuilder.from(TypesPropertyHolder::class.java)

        assertThat(property.children.find { it.name == "doublePrimitive" }).extracting("propertyType").contains(PropertyType.NUMBER)
    }

    @Test
    fun shouldMapDoubleType() {
        val property = propertyBuilder.from(TypesPropertyHolder::class.java)

        assertThat(property.children.find { it.name == "doubleValue" }).extracting("propertyType").contains(PropertyType.NUMBER)
    }
}
